aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/TagApi.java

```diff
@@ -0,0 +1,120 @@
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.ResponseParser;
+import org.jclouds.rest.annotations.Transform;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+
+@Consumes(MediaType.APPLICATION_JSON)
+@RequestFilters(FormSign.class)
+@QueryParams(keys = { "Version", "Format", "SignatureVersion", "ServiceCode", "SignatureMethod" },
+             values = {"{" + Constants.PROPERTY_API_VERSION + "}", "JSON", "1.0", "ecs", "HMAC-SHA1"})
+public interface TagApi {
+
+   @Named("tag:list")
+   @GET
+   @QueryParams(keys = "Action", values = "DescribeTags")
+   @ResponseParser(ParseTags.class)
+   @Fallback(Fallbacks.EmptyIterableWithMarkerOnNotFoundOr404.class)
+   IterableWithMarker<Tag> list(@QueryParam("RegionId") String region, ListTagsOptions options);
+
+   @Named("tag:list")
+   @GET
+   @QueryParams(keys = "Action", values = "DescribeTags")
+   @ResponseParser(ParseTags.class)
+   @Transform(ParseTags.ToPagedIterable.class)
+   @Fallback(Fallbacks.EmptyPagedIterableOnNotFoundOr404.class)
+   PagedIterable<Tag> list(@QueryParam("RegionId") String region);
+
+   @Singleton
+   final class ParseTags extends ParseJson<Tags> {
+
+      @Inject
+      ParseTags(final Json json) {
+         super(json, TypeLiteral.get(Tags.class));
+      }
+
+      static class ToPagedIterable extends Arg0ToPagedIterable<Tag, ToPagedIterable> {
+
+         private final ECSComputeServiceApi api;
+
+         @Inject
+         ToPagedIterable(ECSComputeServiceApi api) {
+            this.api = api;
+         }
+
+         @Override
+         protected Function<Object, IterableWithMarker<Tag>> markerToNextForArg0(final Optional<Object> arg0) {
+            return new Function<Object, IterableWithMarker<Tag>>() {
+               @Override
+               public IterableWithMarker<Tag> apply(Object input) {
+                  String regionId = arg0.get().toString();
+                  ListTagsOptions listTagsOptions = ListTagsOptions.Builder.paginationOptions(PaginationOptions.class.cast(input));
+                  return api.tagApi().list(regionId, listTagsOptions);
+               }
+            };
+         }
+      }
+   }
+
+   @Named("tag:add")
+   @POST
+   @QueryParams(keys = "Action", values = "AddTags")
+   @Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)
+   Request add(@QueryParam("RegionId") String region, @QueryParam("ResourceId") String resourceId,
+                            @QueryParam("ResourceType") String resourceType,
+                            TagOptions tagOptions);
+
+   @Named("tag:remove")
+   @POST
+   @QueryParams(keys = "Action", values = "RemoveTags")
+   @Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)
+   Request remove(@QueryParam("RegionId") String region,
```

Comment:
Does this remove all the tags? Can a single tag be removed?

Reply:
I think so, I'll add a method with optional queryparams to remove one tag only